### PR TITLE
fix(cdk/overlay): backdrop timeouts not being cleared in some cases

### PR DESCRIPTION
**Pull request by crisbeto**
When we start the detach sequence of a backdrop, we bind a `transitionend` event as well as a 500ms timeout in case the event doesn't fire. If the event fires before 500ms, we clear the timeout so users don't have to flush it in tests, however we had the following problems:
1. Transparent backdrops don't have a transition so they were always being cleaned up after the 500ms timeout.
2. We weren't clearing the timeout if the overlay was disposed while the backdrop transition is running. This meant that the timeout would still be in memory, even though the element was removed from the DOM.
3. We had a memory leak where the `click` and `transitionend` events weren't being removed from the backdrop if the overlay is disposed while detaching.

Basically all Material components had one of these issues. These changes resolve the problems by:
1. Clearing the timeout when the overlay is disposed.
2. Setting a 1ms transition on the transparent overlay so its `transitionend` can fire (almost) instantly.
3. Removing the `click` and `transitionend` events when the backdrop is disposed.

### CODE REVIEW

**Review by jelbourn — APPROVED** [review]
lgtm

File: src/cdk/overlay/_index.scss [hunk 1]
```diff
@@ -96,13 +96,8 @@ $backdrop-animation-timing-function: cubic-bezier(0.25, 0.8, 0.25, 1) !default;
   }
 
   .cdk-overlay-transparent-backdrop {
-    // Note: as of Firefox 57, having the backdrop be `background: none` will prevent it from
-    // capturing the user's mouse scroll events. Since we also can't use something like
-    // `rgba(0, 0, 0, 0)`, we work around the inconsistency by not setting the background at
-    // all and using `opacity` to make the element transparent.
-    &, &.cdk-overlay-backdrop-showing {
-      opacity: 0;
```
Comment (mmalerba) [inline src/cdk/overlay/_index.scss:104]: why was the `opacity: 0` removed? this seems to be breaking screenshots in some apps

Comment (crisbeto) [reply src/cdk/overlay/_index.scss:104]: It's been a while since I opened the PR, but I _think_ it was because the `opacity` would actually cause the transition to run, even though it's animation from transparent to transparent.

Comment (mmalerba) [reply src/cdk/overlay/_index.scss:104]: I think there may be something not quite right about it. When I presubmitted this I wound up with some apps that had opaque backdrops covering everything

**Comment by crisbeto** [comment]
I've pushed another change to try and make it easier to land.
